design-documents/service-isolation/caching-layer.md

```diff
@@ -0,0 +1,105 @@
+#### Local caches
+ - Pros: fast, no network traffic associated with retrieving data
+ - Cons: 
+   - shares sources with application
+   - can lead to data inconsistency when multiple instances of application servers are created
+   - the information stored within an individual cache node cannot be shared with other application servers
+   - not follow the 12-factor principles (any data that needs to persist must be stored in a stateful backing service)
+ 
+ ![Local cache](img/local-app-cache.png)
+ 
+#### Remote caches dedicated to each service
+ - Pros: 
+   - scalable, adheres to 12-factor principles
+   - externalize application state
+   - possibility dynamically change the number of application instances
+   - contains only dedicated service data
+ - Cons: 
+   - a slight delay in network calls
+   - difficulties with centralized cache invalidation
+ 
+  ![Remote cache dedicated to each service](img/remote-dedicated-cache.png)
+ 
+#### Remote caches shared across the services
+ - Pros: 
+   - centralized cache invalidation is easy
+   - scalable
+   - adheres to 12-factor principles
+ - Cons: 
+   - cache nodes contain mixed data from different services
+ 
+ ![Remote cache shared across the services](img/remote-cache-cluster.png)
+ 
+Local caches are isolated on the individual nodes and cannot be leveraged within a cluster of application servers.
+Remote distributed caches on the other hand, provide low latency and higher levels of availability, when employed with read replicas and can provide a shared environment for all application servers to utilize the cached data.
+Remote cache dedicated to each service allows to store only data of particular service. 
+But there is no possibility (needs implementation) to flush all cache instances in case of upgrade or other huge changes.
+Remote caches shared across the services are easy to flush centralized, but mix data from different services.
+
+
+#### Cache invalidation
+ - Service should store in cache only own data (DB, configuration, etc.) and be fully responsible for cache invalidation
+ - Service mustn't store in cache data retrieved from other Magento service
+ - In case of "Remote caches dedicated to each service" model service should provide entry point for full cache invalidation
+ - In case of "Remote caches shared across the services" model full cache invalidation can be performed by any service
+
+#### Technologies 
+Today most key/value stores such as Memcached and Redis can store terabytes worth of data. Redis also provides high availability and persistence features. 
+Redis and Memcached offer high performance. Memcached is designed for simplicity while Redis offers a rich set of features that make it effective for a wide range of use cases.
+Unlike Memcached, Redis offers snapshots, replication, transactions, advanced data structures. So I propose to use Redis for application data caching.
+
+### Resolutions
+ - Model "Remote cache dedicated to each service" is most suitable.
```

Comment:
How about a mixed approach? Remote cache makes sense for scaled services. If a service is not impacted with high load, the chances are high that it will have only one instance. Local cache may be more effective in this case.

Reply:
Of course, we should make some performance measurements for our architecture.
But in most cases, the added network overhead associated with a remote cache is of little concern given that a request to it is generally fulfilled in sub-millisecond performance. So to unify cache model for all services I propose to use more advanced option that fit "12-factor app" principles - "Remote cache dedicated to each service"